File: macos/Sources/NoctusCore/Services/CommandRegistry.swift
import Foundation

public final class CommandRegistry {
    private var commands: [String: CommandDefinition] = [:]

    public init() {}

    public func register(_ command: CommandDefinition) {
        precondition(commands[command.id] == nil, "Command '\(command.id)' already registered")
        commands[command.id] = command
    }

    public func getById(_ id: String) -> CommandDefinition? {
        commands[id]
    }

    public func getAll() -> [CommandDefinition] {
        Array(commands.values)
    }

    public func canExecute(_ id: String) -> Bool {
        commands[id]?.canExecute() ?? false
    }

    public func execute(_ id: String) {
        guard let cmd = commands[id], cmd.canExecute() else { return }
        cmd.execute()
    }
}

File: src/Explorer.Core/Services/CommandRegistry.cs
using NoctusExplorer.Core.Models;

namespace NoctusExplorer.Core.Services;

public sealed class CommandRegistry
{
    private readonly Dictionary<string, CommandDefinition> _commands = new(StringComparer.OrdinalIgnoreCase);

    public void Register(CommandDefinition command)
    {
        if (!_commands.TryAdd(command.Id, command))
            throw new ArgumentException($"Command '{command.Id}' is already registered.");
    }

    public CommandDefinition? GetById(string id)
        => _commands.GetValueOrDefault(id);

    public IReadOnlyList<CommandDefinition> GetAll()
        => _commands.Values.ToList().AsReadOnly();

    public bool CanExecute(string id)
        => _commands.TryGetValue(id, out var cmd) && cmd.CanExecute();

    public void Execute(string id)
    {
        if (_commands.TryGetValue(id, out var cmd) && cmd.CanExecute())
            cmd.Execute();
    }
}
